### PR TITLE
refactor(critical-request-chains): split out critical request chains

### PR DESCRIPTION
**Pull request by patrickhulce**
preparing to do some predictive perf analysis and need access to the full request chains, this PR separates out the criticality filter from the wiring of 'initiator'

### CODE REVIEW

File: lighthouse-core/gather/computed/request-chains.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      url: record._url,
+      startTime: record.startTime,
+      endTime: record.endTime,
+      responseReceivedTime: record.responseReceivedTime,
+      transferSize: record.transferSize
+    };
+  }
+
+  /**
+   * Links together network requests (filtered by predicate) with their initiator, building a DAG.
```
Comment (paulirish) [inline lighthouse-core/gather/computed/request-chains.js]: can you add a comment on how the predicate is used?
AFAICT if it returns false on a request, that request and any of its children will be excluded from the graph. Is that right?

File: lighthouse-core/gather/computed/request-chains.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    const requestChains = {};
+    for (const request of requestsToExamine) {
+      // Work back from this request up to the root. If by some weird quirk we are giving request D
+      // here, which has ancestors C, B and A (where A is the root), we will build array [C, B, A]
+      // during this phase.
+      const ancestors = [];
+      let ancestorRequest = request.initiatorRequest();
+      let node = requestChains;
+      while (ancestorRequest) {
+        // If the parent request isn't a high priority request it won't be in the
```
Comment (brendankenny) [inline lighthouse-core/gather/computed/request-chains.js]: should update comment to be more generic

File: lighthouse-core/gather/computed/request-chains.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/gather/computed/request-chains.js]: yep, that's how it was used over in critical. done

File: lighthouse-core/gather/computed/request-chains.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      };
+    }
+
+    return requestChains;
+  }
+
+  /**
+   * @param {!DevtoolsLog} devtoolsLog
+   * @param {!ComputedArtifacts} artifacts
+   * @return {!Promise<!Object>}
```
Comment (brendankenny) [inline lighthouse-core/gather/computed/request-chains.js]: no one's taken this on yet but we probably should :)

Comment (brendankenny) [inline lighthouse-core/gather/computed/request-chains.js]: oh, just kidding, we did it in [`crc-details-renderer`](https://github.com/GoogleChrome/lighthouse/blob/57bcf43a661199d55a7f39ecc008815a2da1fcc6/lighthouse-core/report/v2/renderer/crc-details-renderer.js#L203). Can use `CriticalRequestChainRenderer.CRCNode` but we can make it a generic (not just critical) RequestNode type?

Comment (patrickhulce) [inline lighthouse-core/gather/computed/request-chains.js]: done

File: lighthouse-core/gather/computed/request-chains.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/gather/computed/request-chains.js]: done

**Comment by patrickhulce** [comment]
I don't foresee anyone needing to call `extractChain` with a predicate anytime soon, and if they do they can implement tuple arguments for computed gatherers :) this was mostly a convenience to maintain the existing contract of CriticalRequestChains without causing a bunch of unnecessary noise

**Comment by paulirish** [comment]
Tests are busted in a few ways, it looks like: https://travis-ci.org/GoogleChrome/lighthouse/jobs/248562571

File: lighthouse-core/gather/computed/request-chains.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return 'RequestChains';
+  }
+
+  /**
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {!CriticalRequestChainRenderer.CRCRequest}
+   */
+  static flattenRecord(record) {
+    return {
+      requestId: record.requestId,
```
Comment (paulirish) [inline lighthouse-core/gather/computed/request-chains.js]: let's add priority into this record. (you'll have to normalize between initialpriority and final, i think?

**Comment by patrickhulce** [comment]
based on my digging around last night I'm not going to need this afterall since I'll need to reconstruct dependencies myself for predictive perf, sorry for the wasted cycles :)

**Pull request closed without merging**